src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -930,6 +938,99 @@ void DatabaseCatalog::enqueueDroppedTableCleanup(StorageID table_id, StoragePtr
 }
 
+void DatabaseCatalog::dequeueDroppedTableCleanup(StorageID table_id)
+{
+    String latest_metadata_dropped_path;
+    StorageID dropped_table_id = table_id;
+    TablesMarkedAsDropped::iterator table;
+    {
+        std::lock_guard lock(tables_marked_dropped_mutex);
+        table = tables_marked_dropped.end();
+        time_t latest_drop_time = std::numeric_limits<time_t>::min();
+        for (auto it = tables_marked_dropped.begin(); it != tables_marked_dropped.end(); ++it)
+        {
+            if (it->table_id.uuid == table_id.uuid)
+            {
+                table = it;
+                break;
+            }
+
+            if (it->table_id.database_name == table_id.database_name &&
+                it->table_id.table_name == table_id.table_name &&
+                it->drop_time >= latest_drop_time)
+            {
+                latest_drop_time = it->drop_time;
+                table = it;
+            }
+        }
+
+        if (table == tables_marked_dropped.end())
+            throw Exception(ErrorCodes::UNKNOWN_TABLE,
+                "The drop task of table {} is in progress, has been dropped or the database engine doesn't support it",
+                table_id.getNameForLogs());
+
+        latest_metadata_dropped_path = table->metadata_path;
+        dropped_table_id = table->table_id;
+
+        tables_marked_dropped.erase(table);
+        [[maybe_unused]] auto removed = tables_marked_dropped_ids.erase(dropped_table_id.uuid);
+        assert(removed);
+    }
+
+    LOG_INFO(log, "Trying Undrop table {} from {}", dropped_table_id.getNameForLogs(), latest_metadata_dropped_path);
+    String table_metadata_path = getPathForMetadata(dropped_table_id);
+
+    auto enqueue = [&]()
+    {
+        std::lock_guard lock(tables_marked_dropped_mutex);
+        tables_marked_dropped.emplace_back(*table);
+        tables_marked_dropped_ids.insert(dropped_table_id.uuid);
+        fs::remove(table_metadata_path);
+        CurrentMetrics::add(CurrentMetrics::TablesToDropQueueSize, 1);
+    };
+
+    auto local_context = getContext();
+    ASTPtr ast = DatabaseOnDisk::parseQueryFromMetadata(
+        log, local_context, latest_metadata_dropped_path, /*throw_on_error*/ true, /*remove_empty*/ false);
+    auto * create = typeid_cast<ASTCreateQuery *>(ast.get());
+    if (!create)
+    {
+        enqueue();
+        throw Exception(
+            ErrorCodes::FS_METADATA_ERROR,
+            "Cannot parse metadata of table {} from {}",
+            dropped_table_id.getNameForLogs(),
+            latest_metadata_dropped_path);
+    }
+
+    String data_path = "store/" + getPathForUUID(dropped_table_id.uuid);
+    create->setDatabase(dropped_table_id.database_name);
+    create->setTable(dropped_table_id.table_name);
+    try
+    {
+        auto storage = createTableFromAST(
+            *create,
+            dropped_table_id.getDatabaseName(),
+            data_path,
+            local_context,
+            /* force_restore */ true).second;
```

Comment:
It's unsafe to create another instance of the table while the previous one exists. See `DatabaseAtomic::waitDetachedTableNotInUse`.